src/processors/rendering.rs

```diff
@@ -0,0 +1,427 @@
+//! Default rendering processor types.
+
+extern crate cgmath;
+
+use ecs::{Processor, RunArg, Join, Component, VecStorage, Entity};
+use context::Context;
+use std::sync::{Mutex, Arc};
+use renderer;
+use renderer::Layer;
+use std::collections::HashSet;
+use config::Element;
+use std::path::Path;
+
+config!(
+    /// A config required to create a rendering processor.
+    struct RendererConfig {
+        // Forward or Deferred
+        pub pipeline: String = "Forward".to_string(),
+        // Flat or Shaded
+        pub shading: String = "Flat".to_string(),
+        pub clear_color: [f32; 4] = [0., 0., 0., 1.],
+    }
+);
+
+/// A rendering processor struct.
+pub struct RenderingProcessor {
+    active_camera: Option<Entity>,
+}
+
+const ACTIVE_CAMERA_NAME: &'static str = "main";
+const ACTIVE_SCENE_NAME: &'static str = "main";
+
+fn forward_flat(clear_color: [f32; 4]) -> Vec<Layer> {
+    use renderer::pass::*;
+
+    vec![
+        Layer::new("main",
+            vec![
+                Clear::new(clear_color),
+                DrawFlat::new(ACTIVE_CAMERA_NAME, ACTIVE_SCENE_NAME),
+            ]
+        ),
+    ]
+}
+
+fn forward_shaded(clear_color: [f32; 4]) -> Vec<Layer> {
+    use renderer::pass::*;
+
+    vec![
+        Layer::new("main",
+            vec![
+                Clear::new(clear_color),
+                DrawShaded::new(ACTIVE_CAMERA_NAME, ACTIVE_SCENE_NAME),
+            ]
+        ),
+    ]
+}
+
+fn layer_gbuffer(clear_color: [f32; 4]) -> Layer {
+    use renderer::pass::*;
+
+    Layer::new("gbuffer",
+        vec![
+            Clear::new(clear_color),
+            DrawFlat::new(ACTIVE_CAMERA_NAME, ACTIVE_SCENE_NAME),
+        ]
+    )
+}
+
+fn deferred_flat(clear_color: [f32; 4]) -> Vec<Layer> {
+    use renderer::pass::*;
+
+    vec![
+        layer_gbuffer(clear_color),
+        Layer::new("main",
+            vec![
+                BlitLayer::new("gbuffer", "ka"),
+            ]
+        ),
+    ]
+}
+
+fn deferred_shaded(clear_color: [f32; 4]) -> Vec<Layer> {
+    use renderer::pass::*;
+
+    vec![
+        layer_gbuffer(clear_color),
+        Layer::new("main",
+            vec![
+                BlitLayer::new("gbuffer", "ka"),
+                Lighting::new(ACTIVE_CAMERA_NAME, "gbuffer", ACTIVE_SCENE_NAME)
+            ]
+        ),
+    ]
+}
+
+impl RenderingProcessor {
+    pub fn new(renderer_config: RendererConfig, context: &mut Context) -> RenderingProcessor {
+        let clear_color = renderer_config.clear_color;
+        let pipeline = match renderer_config.pipeline.as_str() {
```

Comment:
Ni: this might be tidier if you use a single match.

```rust
match (renderer_config.pipeline.as_str(), renderer_config.shading.as_str()) {
    ("Forward", "Flat") => forward_flat(clear_color),
    ("Forward", "Shaded") => forward_shaded(clear_color),
    ("Deferred", "Flat") => deferred_flat(clear_color),
    ("Deferred", "Shaded") => deferred_shaded(clear_color),
    _ => panic!("Error: Can't provide rendering pipeline requested in renderer_config, \
                 renderer_config.shading field is invalid.")
}